File: backend/app/schemas/policy.py
"""Schemas for policies and policy mappings."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class PolicyBase(BaseModel):
    """Base schema for Policy."""
    name: str = Field(..., max_length=255)
    description: str | None = None
    version: str | None = Field(default=None, max_length=50)
    owner: str | None = Field(default=None, max_length=255)


class PolicyCreate(PolicyBase):
    """Schema for creating a Policy."""
    assessment_id: UUID
    file_path: str | None = None
    content_text: str | None = None


class PolicyUpdate(BaseModel):
    """Schema for updating a Policy."""
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None
    version: str | None = Field(default=None, max_length=50)
    owner: str | None = Field(default=None, max_length=255)


class PolicyResponse(PolicyBase):
    """Schema for Policy response."""
    id: UUID
    assessment_id: UUID
    file_path: str | None = None
    content_text: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PolicyMappingBase(BaseModel):
    """Base schema for Policy Mapping."""
    policy_id: UUID
    subcategory_id: UUID
    confidence_score: float | None = Field(default=None, ge=0.0, le=1.0)


class PolicyMappingCreate(PolicyMappingBase):
    """Schema for creating a Policy Mapping."""
    pass


class PolicyMappingResponse(PolicyMappingBase):
    """Schema for Policy Mapping response."""
    id: UUID
    is_approved: bool
    approved_by_id: UUID | None = None
    approved_at: datetime | None = None
    created_at: datetime
    subcategory_code: str | None = None
    policy_name: str | None = None

    model_config = {"from_attributes": True}


class PolicyUploadResponse(BaseModel):
    """Response for policy file upload."""
    policy: PolicyResponse
    text_extracted: bool
    text_length: int | None = None
    extraction_error: str | None = None

File: frontend/lib/types/policy.ts
import { UUID } from './common';

export interface Policy {
  id: UUID;
  assessment_id: UUID;
  name: string;
  description: string | null;
  version: string | null;
  owner: string | null;
  file_path: string | null;
  content_text: string | null;
  created_at: string;
  updated_at: string;
}

export interface PolicyUploadResponse {
  policy: Policy;
  text_extracted: boolean;
  text_length: number | null;
  extraction_error: string | null;
}

export interface PolicyMapping {
  id: UUID;
  policy_id: UUID;
  subcategory_id: UUID;
  confidence_score: number | null;
  is_approved: boolean;
  approved_by_id: UUID | null;
  approved_at: string | null;
  created_at: string;
  subcategory_code: string | null;
  policy_name: string | null;
}
